### PR TITLE
[USD Price] SWR-ise the USD fetching logic

### PR DESCRIPTION
**Pull request by W3stside**
# Summary

Attempts to SWR-ise the USD querying logic

As a bonus - stops 2 faulty calls to quote happening on initial load - app was querying "USDC" in price quote

### CODE REVIEW

**Comment by elena-zh** [comment]
Could you please provide the PR preview link?

**Comment by elena-zh** [comment]
Hey @W3stside , is it OK that even if I select ETH-WETH or WXDAI-XDAI token pairs, quote requests for WETH-USDC, WXDAI-USDXC are still running in the console for Rinkeby and GC networks? 
image

**Comment by elena-zh** [comment]
Also, I noticed, that USD estimation for ETH token  can disappear when swap tokens in the From/TO fields
image
image

File: src/custom/hooks/useUSDCPrice/index.ts [hunk 1]
```diff
@@ -84,66 +83,78 @@ export default function useUSDCPrice(currency?: Currency) {
...
+          console.debug(
+            '[useBestUSDCPrice] Best USDC price amount',
+            price.toSignificant(12),
+            price.invert().toSignificant(12)
+          )
+        }
+
+        return setBestUsdPrice(price)
+      } catch (err) {
+        console.error('[useBestUSDCPrice] Error getting best price', err)
```
Comment (alfetopito) [inline src/custom/hooks/useUSDCPrice/index.ts:150]: Nitpick: update the comments to reflect new function name

File: src/custom/state/price/updater.ts [hunk 2]
```diff
@@ -151,13 +152,26 @@ export default function FeesUpdater(): null {
   const isOnline = useIsOnline()
   const { validTo } = useOrderValidTo()
 
+  // prevents things like "USDC" being used as an address
```
Comment (alfetopito) [inline src/custom/state/price/updater.ts:155]: Ah yes, `USDC` token is default now. If I recall correctly it'll be replaced by the address eventually then this should just work

**Comment by alfetopito** [comment]
Not sure the caching is working as expected, at least not how I would expect.
Using Elena's use case (switching between sell/buy order), every swap sends a new `quote` api endpoint request.
I would expect it to not be sent unless caching is expired.

https://user-images.githubusercontent.com/43217/168611556-6bdf871b-7ecf-45a6-9c31-5c762e48dae2.mov

BTW, what is the cache expiring time?

**Comment by W3stside** [comment]
> Not sure the caching is working as expected, at least not how I would expect. Using Elena's use case (switching between sell/buy order), every swap sends a new `quote` api endpoint request. I would expect it to not be sent unless caching is expired.
> 
>  Screen.Recording.2022-05-16.at.15.05.54.mov 
> BTW, what is the cache expiring time?

Well any new swap = different swap params = invalidated cache, or did u mean sth else?

Not sure the cache time, think it's 20s (20000) or default

File: src/custom/hooks/useUSDCPrice/index.ts [hunk 3]
```diff
@@ -84,66 +83,78 @@ export default function useUSDCPrice(currency?: Currency) {
...
       fromDecimals: sellTokenDecimals,
       toDecimals: stablecoin.decimals,
       userAddress: account,
       // we dont care about validTo here, just use max
       validTo: MAX_VALID_TO_EPOCH,
     }
+  }, [account, baseAmountRaw, sellTokenAddress, sellTokenDecimals, stablecoin, supportedChain])
+
+  // SWR cache cow usd requests
+  const { data: priceResponse, error: errorResponse } = useSWR<CancelableResult<string | null> | null>(
```
Comment (anxolin) [inline src/custom/hooks/useUSDCPrice/index.ts:109]: how often this cached result will be revalidated? when does it become stale?

File: src/custom/hooks/useUSDCPrice/index.ts [hunk 4]
```diff
@@ -194,60 +205,69 @@ export function useCoingeckoUsdPrice(currency?: Currency) {
 
   const tokenAddress = currencyRef.current ? currencyId(currencyRef.current) : undefined
 
+  const chainIdSupported = supportedChainId(chainId)
+  // SWR cache cg usd requests
+  const { data: priceResponse, error: errorResponse } = useSWR<PriceInformation | null>(
+    ['coingeckoUsdPrice', chainIdSupported, tokenAddress],
```
Comment (anxolin) [inline src/custom/hooks/useUSDCPrice/index.ts:217]: nit: why not calling the cache exactly the same name as the function `getUSDPriceQuote`?

**Review by anxolin — COMMENTED** [review]
Nice initiative. 

My only problem with this PR is where you are adding caches. 

In principle, the cache should be added in quering of the data.  Adding it there feels like mixing logic with cache strategies.  The logics for the strategy quering should be fast. Why not just wrapping the API calls with a simple custom hook that uses SWR

**Comment by elena-zh** [comment]
USD estimation for ETH token work fine now, but the issue reported in https://github.com/cowprotocol/cowswap/pull/539#issuecomment-1122222665 is still here

**Comment by alfetopito** [comment]
> > Not sure the caching is working as expected, at least not how I would expect. Using Elena's use case (switching between sell/buy order), every swap sends a new `quote` api endpoint request. I would expect it to not be sent unless caching is expired.
> > Screen.Recording.2022-05-16.at.15.05.54.mov
> > BTW, what is the cache expiring time?
> 
> Well here you're looking at price quotes. We're talking USD estimation, which isn't being called AFAICT. also any new swap = different swap params = invalidated cache, or did u mean sth else?
> 
> Not sure the cache time, think it's 2s (2000) which is default

Ah sorry, my mistake.
Was watching for our price quotes not the USD quotes - which is what you changed 🤦

**Comment by alfetopito** [comment]
Filtering for coingecko requests though, they all seem to be failing due to CORS issues
Screen Shot 2022-05-20 at 13 37 23

**Review by alfetopito — APPROVED** [review]

**Comment by alfetopito** [comment]
> Filtering for coingecko requests though, they all seem to be failing due to CORS issues Screen Shot 2022-05-20 at 13 37 23

Seemed to be a temporary fluke locally. All is well now

File: src/custom/hooks/useUSDCPrice/index.ts (hunk 4 above)
Comment (W3stside) [reply src/custom/hooks/useUSDCPrice/index.ts:217]: done

**Comment by W3stside** [comment]
added SWR options to constants:
```ts
const SWR_OPTIONS = {
  refreshInterval: ms`30s`,
  dedupingInterval: ms`10s`,
  // don't revalidate data on focus, can cause too many re-renders
  // see https://koba04.medium.com/revalidating-options-of-swr-4d9f08bee813
  revalidateOnFocus: false,
}
```
- refreshes the data every 30 seconds
- doesn't revalidate on focus as that can cause too many revalidations
- dedupes requests within a 10s window

@anxolin @alfetopito thoughts?

File: src/custom/hooks/useUSDCPrice/index.ts (hunk 3 above)
Comment (W3stside) [reply src/custom/hooks/useUSDCPrice/index.ts]: it's currently revalidated using default options, as shown:
image

i added custom options that are in a comment below

**Comment by alfetopito** [comment]
> added SWR options to constants:
> 
> ```ts
> const SWR_OPTIONS = {
>   refreshInterval: ms`30s`,
>   dedupingInterval: ms`10s`,
>   // don't revalidate data on focus, can cause too many re-renders
>   // see https://koba04.medium.com/revalidating-options-of-swr-4d9f08bee813
>   revalidateOnFocus: false,
> }
> ```
> 
>     * refreshes the data every 30 seconds
> 
>     * doesn't revalidate on focus as that can cause too many revalidations
> 
>     * dedupes requests within a 10s window
> 
> 
> @anxolin @alfetopito thoughts?

Looks good.

**Comment by W3stside** [comment]
ok im merging this

**Pull request merged**